lib/ansible/module_utils/common/arg_spec.py

```diff
@@ -5,130 +5,265 @@
         self._unsupported_parameters = set()
-
-    @property
-    def error_messages(self):
-        return self._error_messages
+        self._validated_parameters = deepcopy(parameters)
+        self.deprecations = []
+        self.errors = AnsibleValidationErrorMultiple()
+        self.warnings = []
 
     @property
     def validated_parameters(self):
         return self._validated_parameters
 
-    def _add_error(self, error):
-        if isinstance(error, string_types):
-            self._error_messages.append(error)
-        elif isinstance(error, Sequence):
-            self._error_messages.extend(error)
-        else:
-            raise ValueError('Error messages must be a string or sequence not a %s' % type(error))
+    @property
+    def unsupported_parameters(self):
+        return self._unsupported_parameters
+
+    @property
+    def error_messages(self):
+        return self.errors.messages
+
+
+class ArgumentSpecValidator:
+    """Argument spec validation class
+
+    Creates a validator based on the ``argument_spec`` that can be used to
+    validate a number of parameters using the ``validate()`` method.
+
+    :param argument_spec: Specification of valid parameters and their type. May
+        include nested argument specs.
+    :type argument_spec: dict
+
+    :param mutually_exclusive: List or list of lists of terms that should not
+        be provided together.
+    :type mutually_exclusive: list, optional
+
+    :param required_together: List of lists of terms that are required together.
+    :type required_together: list, optional
 
-    def _sanitize_error_messages(self):
-        self._error_messages = remove_values(self._error_messages, self._no_log_values)
+    :param required_one_of: List of lists of terms, one of which in each list
+        is required.
+    :type required_one_of: list, optional
 
-    def validate(self, *args, **kwargs):
-        """Validate module parameters against argument spec.
+    :param required_if: List of lists of ``[parameter, value, [parameters]]`` where
+        one of [parameters] is required if ``parameter`` == ``value``.
+    :type required_if: list, optional
+
+    :param required_by: Dictionary of parameter names that contain a list of
+        parameters required by each key in the dictionary.
+    :type required_by: dict, optional
+    """
+
+    def __init__(self, argument_spec,
+                 mutually_exclusive=None,
+                 required_together=None,
+                 required_one_of=None,
+                 required_if=None,
+                 required_by=None,
+                 ):
+
+        self._valid_parameter_names = set()
+        self._mutually_exclusive = mutually_exclusive
+        self._required_together = required_together
+        self._required_one_of = required_one_of
+        self._required_if = required_if
+        self._required_by = required_by
+        self.argument_spec = argument_spec
+
+    @property
+    def valid_parameter_names(self):
```

Comment:
I agree with the above three statements. I _really_ dislike the duplication of sub-spec validation in its own explicit step, but refactoring that was too much while moving all the other validation code. I'll put a FIXME in there since it really should be integrated into the existing validation rather than its own thing.

> we'd almost certainly have to restructure it later to create actual warning objects if we have more than one

I was thinking custom objects would make sense as well. It's unfortunate that the primary use case right now is generating an error message string, so that's clouding my thinking.

This is more or less the same problem with `valid_parameter_names` as well. But I need some time to think through how to make this more generally useful. I will mark these as private for now.